File: apple/MuCore/Tests/MuCoreTests/ErrFixtureTest.swift
import XCTest
@testable import MuCore

/// 錯誤語意契約測試：跑 contract/fixtures/err_cases/。
/// RetryPolicy 輸出必須與 Python 參考實作 byte-identical（provider.md §2.1）。
final class ErrFixtureTest: XCTestCase {

    func testAllErrFixtureCasesMatchByteForByte() throws {
        let casesDir = try XCTUnwrap(findDir("contract/fixtures/err_cases"),
                                     "err_cases not found")
        let names = try XCTUnwrap(
            try FileManager.default.contentsOfDirectory(atPath: casesDir.path)
                .filter { name -> Bool in
                    var isDir: ObjCBool = false
                    let ok = FileManager.default.fileExists(
                        atPath: casesDir.appendingPathComponent(name).path, isDirectory: &isDir)
                    return ok && isDir.boolValue
                }
                .sorted())
        XCTAssertFalse(names.isEmpty)

        var failures: [String] = []
        for name in names {
            let caseDir = casesDir.appendingPathComponent(name)
            let expected = try String(contentsOf: caseDir.appendingPathComponent("expected.json"),
                                      encoding: .utf8)
            let script = try XCTUnwrap(
                JSONSerialization.jsonObject(
                    with: try Data(contentsOf: caseDir.appendingPathComponent("script.json")))
                as? [String: Any])
            var out: [CanonicalJson.JSONValue] = []
            for e in try XCTUnwrap(script["entries"] as? [[String: Any]]) {
                XCTAssertEqual(e["type"] as? String, "retry", "unknown entry type in case [\(name)]")
                var queue = (e["script"] as? [String]) ?? []
                let o = RetryPolicy.run(
                    op: {
                        guard let s = queue.first else { return nil }
                        queue.removeFirst()
                        return s == "ok" ? nil : RetryPolicy.kind(from: s)
                    },
                    onReauth: {},
                    sleep: { _ in })
                out.append(.object([
                    ("reauths", .int(o.reauths)),
                    ("result", .string(o.result)),
                    ("sleeps", .array(o.sleeps.map { .int($0) })),
                ]))
            }
            let actual = CanonicalJson.render(.array(out))
            if expected != actual {
                failures.append("case [\(name)]\n--- expected ---\n\(expected)\n" +
                    "--- actual ---\n\(actual)")
            }
        }
        if !failures.isEmpty {
            XCTFail("\(failures.count)/\(names.count) err cases drifted:\n\n" +
                failures.joined(separator: "\n\n========\n\n"))
        }
    }

    private func findDir(_ rel: String) -> URL? {
        var dir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        while true {
            let c = dir.appendingPathComponent(rel)
            if FileManager.default.fileExists(atPath: c.path) { return c }
            if dir.path == "/" { return nil }
            dir = dir.deletingLastPathComponent()
        }
    }
}

File: android/core/src/test/kotlin/crate/core/ErrFixtureTest.kt
package crate.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File
import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.test.fail

/**
 * 錯誤語意契約測試：跑 contract/fixtures/err_cases/。
 * RetryPolicy 輸出必須與 Python 參考實作 byte-identical（provider.md §2.1）。
 */
class ErrFixtureTest {

    @Test
    fun `all err fixture cases match expected json byte for byte`() {
        val casesDir = findDir("contract/fixtures/err_cases") ?: fail("err_cases not found")
        val names = casesDir.listFiles()?.filter { it.isDirectory }?.map { it.name }?.sorted()
            ?: fail("cannot list $casesDir")
        assertTrue(names.isNotEmpty())

        val failures = ArrayList<String>()
        for (name in names) {
            val caseDir = File(casesDir, name)
            val expected = File(caseDir, "expected.json").readBytes()
            val script = Json.parseToJsonElement(
                File(caseDir, "script.json").readText()).jsonObject
            val out = ArrayList<Map<String, Any?>>()
            for (e in script["entries"]!!.jsonArray.map { it.jsonObject }) {
                if (e["type"]!!.jsonPrimitive.content != "retry") {
                    fail("unknown entry type in case [$name]")
                }
                val queue = e["script"]!!.jsonArray
                    .map { it.jsonPrimitive.content }.toMutableList()
                val o = RetryPolicy.run(
                    op = {
                        when (val s = queue.removeFirstOrNull()) {
                            null, "ok" -> null
                            else -> RetryPolicy.kindFromJson(s)
                        }
                    },
                    onReauth = { },
                    sleep = { /* 收集在 Outcome.sleeps */ },
                )
                out.add(linkedMapOf<String, Any?>(
                    "reauths" to o.reauths, "result" to o.result, "sleeps" to o.sleeps))
            }
            val actual = CanonicalJson.render(out).toByteArray(Charsets.UTF_8)
            if (!expected.contentEquals(actual)) {
                failures.add("case [$name]\n--- expected ---\n${expected.decodeToString()}\n" +
                    "--- actual ---\n${actual.decodeToString()}")
            }
        }
        if (failures.isNotEmpty()) {
            fail("${failures.size}/${names.size} err cases drifted:\n\n" +
                failures.joinToString("\n\n========\n\n"))
        }
    }

    private fun findDir(rel: String): File? {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (dir != null) {
            val c = File(dir, rel)
            if (c.isDirectory) return c
            dir = dir.parentFile
        }
        return null
    }
}
